codex-rs/execpolicy2/src/rule.rs

```diff
@@ -0,0 +1,220 @@
+            Self::Single(expected) => expected == token,
+            Self::Alts(alternatives) => alternatives.iter().any(|alt| alt == token),
+        }
+    }
+
+    pub fn alternatives(&self) -> Vec<String> {
+        match self {
+            Self::Single(expected) => vec![expected.clone()],
+            Self::Alts(alternatives) => alternatives.clone(),
+        }
+    }
+}
+
+impl std::fmt::Display for PatternToken {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(render_pattern_token(self).as_str())
+    }
+}
+
+/// Prefix matcher for commands with support for alternative match tokens.
+/// First token is fixed since we key by the first token in policy.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct PrefixPattern {
+    pub first: String,
+    pub rest: Vec<PatternToken>,
+}
+
+impl PrefixPattern {
+    pub fn len(&self) -> usize {
+        self.rest.len() + 1
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn matches_prefix(&self, cmd: &[String]) -> Option<Vec<String>> {
+        if cmd.len() < self.len() || cmd[0] != self.first {
+            return None;
+        }
+
+        for (pattern_token, cmd_token) in self.rest.iter().zip(&cmd[1..self.len()]) {
+            if !pattern_token.matches(cmd_token) {
+                return None;
+            }
+        }
+
+        Some(cmd[..self.len()].to_vec())
+    }
+}
+
+impl std::fmt::Display for PrefixPattern {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "[{}]", render_pattern(self))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum Rule {
+    Prefix(PrefixRule),
+}
+
+impl std::fmt::Display for Rule {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Prefix(rule) => write!(
+                f,
+                "prefix_rule(pattern = {}, decision = {})",
+                rule.pattern, rule.decision
+            ),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct PrefixRule {
+    pub pattern: PrefixPattern,
+    pub decision: Decision,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum RuleMatch {
+    PrefixRuleMatch {
+        matched_prefix: Vec<String>,
+        decision: Decision,
+    },
+}
+
+impl RuleMatch {
+    pub fn decision(&self) -> Decision {
+        match self {
+            Self::PrefixRuleMatch { decision, .. } => *decision,
+        }
+    }
+}
+
+impl std::fmt::Display for RuleMatch {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::PrefixRuleMatch {
+                matched_prefix,
+                decision,
+            } => write!(
+                f,
+                "prefixRuleMatch {{ matchedPrefix: {matched_prefix:?}, decision: {decision} }}"
+            ),
+        }
+    }
+}
+
+impl Rule {
```

Comment:
`impl` should be after the declaration for readability: please update throughout